File: app/src/main/java/com/example/myapplication/data/CoachExplanationClient.kt
package com.example.myapplication.data

import com.example.myapplication.core.adaptation.AdaptationKind
import java.util.concurrent.TimeUnit
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

interface CoachExplanationClient {
    suspend fun explainDecision(
        kind: AdaptationKind,
        reasonVi: String,
        beforeValue: String,
        afterValue: String,
    ): String?
}

class OkHttpCoachExplanationClient(
    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .build(),
    private val endpointUrl: String = "http://10.0.2.2:3000/api/explain-decision",
) : CoachExplanationClient {
    override suspend fun explainDecision(
        kind: AdaptationKind,
        reasonVi: String,
        beforeValue: String,
        afterValue: String,
    ): String? = withContext(Dispatchers.IO) {
        val payload = buildJsonObject {
            put("kind", kind.name)
            put("reasonVi", reasonVi)
            put("beforeValue", beforeValue)
            put("afterValue", afterValue)
        }.toString()

        val request = Request.Builder()
            .url(endpointUrl)
            .post(payload.toRequestBody("application/json".toMediaType()))
            .build()

        try {
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) return@withContext null
                val bodyString = response.body?.string() ?: return@withContext null
                val element = Json.parseToJsonElement(bodyString).jsonObject
                val explanation = element["explanation"]?.jsonPrimitive?.content
                if (explanation.isNullOrBlank()) null else explanation
            }
        } catch (e: Exception) {
            null // Fallback locally on timeout or network error
        }
    }
}

File: flutter/lib/data/remote/coach_explanation_client.dart
import 'package:dio/dio.dart';
import '../../core/model/adaptation_models.dart';
import 'backend_config.dart';

abstract class CoachExplanationClient {
  Future<String?> explainDecision({
    required AdaptationKind kind,
    required String reasonVi,
    required String beforeValue,
    required String afterValue,
  });
}

class DioCoachExplanationClient implements CoachExplanationClient {
  final Dio _dio;
  final String? Function() _endpointProvider;

  DioCoachExplanationClient({
    Dio? dio,
    String? Function()? endpointProvider,
  })  : _dio = dio ?? Dio(BaseOptions(
          connectTimeout: const Duration(seconds: 10),
          receiveTimeout: const Duration(seconds: 10),
        )),
        _endpointProvider = endpointProvider ??
            (() => BackendConfig.baseUrl != null
                ? '${BackendConfig.baseUrl}/api/coach/decision-explanations'
                : null);

  @override
  Future<String?> explainDecision({
    required AdaptationKind kind,
    required String reasonVi,
    required String beforeValue,
    required String afterValue,
  }) async {
    final endpointUrl = _endpointProvider();
    if (endpointUrl == null) return null;

    try {
      final response = await _dio.post(
        endpointUrl,
        data: {
          'kind': kind.name.toUpperCase(), // Khớp với enum.name của Kotlin
          'reasonVi': reasonVi,
          'beforeValue': beforeValue,
          'afterValue': afterValue,
        },
        options: Options(
          contentType: Headers.jsonContentType,
        ),
      );

      if (response.statusCode != 200) return null;

      final data = response.data;
      if (data is Map<String, dynamic>) {
        final explanation = data['explanation'] as String?;
        if (explanation != null && explanation.trim().isNotEmpty) {
          return explanation;
        }
      }
      return null;
    } catch (_) {
      return null; // Fallback locally on timeout or network error
    }
  }
}
